### ISSUE TITLE
Empty datatable using the service implementation 

### ISSUE THREAD

**Issue by swerwajod** [yajra/laravel-datatables#1413]
### Data table shows header but is empty, no HTML rows are created in the table.
The implementation is Laravel 5.5 on a Homestead box on windows 10.  This all works well as I have been using the application I built for some time without issues.  I decided to add the laravel-datatables as a service to make improvements to the table in the application.
I installed your datatables v8 install through 
`composer require yajra/laravel-datatables:^1.0` 
the service providers were explicitly added to the app.php
```
Yajra\DataTables\DataTablesServiceProvider::class,
Yajra\DataTables\HtmlServiceProvider::class,
Yajra\DataTables\ButtonsServiceProvider::class,
Yajra\Datatables\FractalServiceProvider::class,
```
I used the command `php artisan datatables:make Users` pasted here
```php
<?php
// App\Datatables\UsersDataTables

namespace App\DataTables;

use App\User;
use Yajra\DataTables\Services\DataTable;
use Yajra\DataTables\EloquentDataTable;

class UsersDataTable extends DataTable
{
    /**
     * Build DataTable class.
     *
     * @param mixed $query Results from query() method.
     * @return \Yajra\DataTables\DataTableAbstract
     */
    public function dataTable($query)
    {
        $dataTable = new EloquentDataTable($query);

        return $dataTable->addColumn('action', 'users.action');
    }

    /**
     * Get query source of dataTable.
     *
     * @param \App\User $model
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function query(User $model)
    {
        return $model->newQuery()->select($this->getColumns());
    }

    /**
     * Optional method if you want to use html builder.
     *
     * @return \Yajra\DataTables\Html\Builder
     */
    public function html()
    {
        return $this->builder()
                    ->columns($this->getColumns())
                    ->minifiedAjax()
                    ->addAction(['width' => '80px'])
                    ->parameters([
                        'dom'     => 'Bfrtip',
                        'order'   => [[0, 'desc']],
                        'buttons' => [
                            'create',
                            'export',
                            'print',
                            'reset',
                            'reload',
                        ],
                    ]);
    }

    /**
     * Get columns.
     *
     * @return array
     */
    protected function getColumns()
    {
        return [
            'username',
            'name',
            'email',
            'telephonenumber',
            'displayname'
        ];
    }

    /**
     * Get filename for export.
     *
     * @return string
     */
    protected function filename()
    {
        return 'users_' . time();
    }
}
```
My route in web.php is
```php
Route::get('/users', 'UsersController@index');
```
My UsersContoller.php contains the index method
```php
    public function index(UsersDataTable $dataTable)
    {
        return $dataTable->render('users.index');
    }
```
My index.blade.php is
```php
@extends('layouts.app')

@section('title', '| Users')

@section('content')
    {!! $dataTable->table() !!}
@endsection

@push('scripts')
    <link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.0.3/css/buttons.dataTables.min.css">
    <script src="https://cdn.datatables.net/buttons/1.0.3/js/dataTables.buttons.min.js"></script>
    <script src="/vendor/datatables/buttons.server-side.js"></script>
    {!! $dataTable->scripts() !!}
@endpush
```
With the layouts.app file containing ```    @stack('scripts')``` in the html head
the page source shows the included scripts in the head and the table with only the column header row.

I don't know at this point what is missing. If I tell the controller to just return a collection of the data I do get a collection of the data.
```php
return \DataTables::eloquent(User::query())->make(true);
```
So the eloquent part with a direct query seems to work. Just the UsersDataTable is not sending any more than the column headers to the HTML.
I reviewed the demos and similar issues but found no answer.  The demos all seem to be for earlier versions and the issues for missing data are either different reasons or no solution is presented.  I would really like to use the service method of you datatables.  I'm betting this is something simple that is missing.

### System details

- Operating System - homestead box on windows 10
- PHP Version - 7.1
- Laravel Version - 5.5
- Laravel-Datatables Version - 8.0 installed by complete installer 1.0

**Comment by yajra** [comment]
The php code looks good. Is there any errors on browser's console? Might be an issue on javascripts. Is the main `dataTable.js` library included on your master layout?

**Comment by swerwajod** [comment]
Console errror is

> dataTables.buttons.min.js:32 Uncaught TypeError: Cannot read property 'Buttons' of undefined
    at dataTables.buttons.min.js:32
    at dataTables.buttons.min.js:32
(anonymous) @ dataTables.buttons.min.js:32
(anonymous) @ dataTables.buttons.min.js:32

The scripts included in the page source is 
```HTML
<!-- Styles -->
 <link href="https://dash.app/css/app.css" rel="stylesheet">
 <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
 <link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.0.3/css/buttons.dataTables.min.css">
 <link rel="stylesheet" href="https://cdn.datatables.net/1.10.16/css/jquery.dataTables.min.css">
 <script src="https://cdn.datatables.net/buttons/1.0.3/js/dataTables.buttons.min.js"></script>
 <script src="https://cdn.datatables.net/1.10.16/js/jquery.dataTables.min.js"></script>
 <script src="/vendor/datatables/buttons.server-side.js"></script>
 <script type="text/javascript">(function(window,$){window.LaravelDataTables=window.LaravelDataTables\|\|{};window.LaravelDataTables["dataTableBuilder"]=$("#dataTableBuilder").DataTable({"serverSide":true,"processing":true,"ajax":{"url":"","type":"GET","data":function(data) {
 for (var i = 0, len = data.columns.length; i < len; i++) {
 if (!data.columns[i].search.value) delete data.columns[i].search;
 if (data.columns[i].searchable === true) delete data.columns[i].searchable;
 if (data.columns[i].orderable === true) delete data.columns[i].orderable;
 if (data.columns[i].data === data.columns[i].name) delete data.columns[i].name;
 }
 delete data.search.regex;}},"columns":[{"name":"id","data":"id","title":"Id","orderable":true,"searchable":true},{"name":"username","data":"username","title":"Username","orderable":true,"searchable":true},{"name":"name","data":"name","title":"Name","orderable":true,"searchable":true},{"name":"email","data":"email","title":"Email","orderable":true,"searchable":true},{"name":"telephonenumber","data":"telephonenumber","title":"Telephonenumber","orderable":true,"searchable":true},{"name":"displayname","data":"displayname","title":"Displayname","orderable":true,"searchable":true},{"name":"created_at","data":"created_at","title":"Created At","orderable":true,"searchable":true},{"name":"updated_at","data":"updated_at","title":"Updated At","orderable":true,"searchable":true},{"defaultContent":"","data":"action","name":"action","title":"Action","render":null,"orderable":false,"searchable":false,"width":"80px"}],"dom":"Bfrtip","order":[[0,"desc"]],"buttons":["create","export","print","reset","reload"]});})(window,jQuery);
 </script>
```

**Comment by yajra** [comment]
Try using the latest `buttons.js` version. I think the latest version is around 1.4.2 and 1.0.3 is a bit old.

**Comment by swerwajod** [comment]
```HTML
<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
 <link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.4.2/css/buttons.dataTables.min.css">
 <link rel="stylesheet" href="https://cdn.datatables.net/1.10.16/css/jquery.dataTables.min.css">
 <script src="https://cdn.datatables.net/buttons/1.4.2/js/dataTables.buttons.min.js"></script>
 <script src="https://cdn.datatables.net/1.10.16/js/jquery.dataTables.min.js"></script>
 <script src="/vendor/datatables/buttons.server-side.js"></script>
```
Yes just tried that after last reply. now the console error reads.

> dataTables.buttons.min.js:5 Uncaught TypeError: Cannot read property 'ext' of undefined
    at dataTables.buttons.min.js:5
    at dataTables.buttons.min.js:5
    at dataTables.buttons.min.js:5

**Comment by jamesriady** [comment]
try using only this:
`<link href="//netdna.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">`
`<link rel="stylesheet" href="//cdn.datatables.net/1.10.7/css/jquery.dataTables.min.css">`
`<script src="//cdn.datatables.net/1.10.7/js/jquery.dataTables.min.js"></script>`

Hope this can help you..

**Comment by swerwajod** [comment]
@jamesriady Thank you for your input. I did try it but it provided the same result. Still only the table header row.

@everyone
The console error is now gone I needed to place the jquery first.
Here is the current html output. (I removed the menus for clarity.
```HTML
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <!-- CSRF Token -->

    <title>LJT DASH</title>

    <!-- Styles -->
    <link href="https://netdna.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <link href="https://dash.app/css/app.css" rel="stylesheet">
    <link rel="stylesheet" href="https://cdn.datatables.net/1.10.16/css/jquery.dataTables.min.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.2.1/jquery.min.js"></script>
    <script src="https://cdn.datatables.net/1.10.16/js/jquery.dataTables.min.js"></script>
    <!-- from dataTables push-->
        <link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.4.2/css/buttons.dataTables.min.css">
    <script src="https://cdn.datatables.net/buttons/1.4.2/js/dataTables.buttons.min.js"></script>
    <script src="/vendor/datatables/buttons.server-side.js"></script>
    <script type="text/javascript">(function(window,$){window.LaravelDataTables=window.LaravelDataTables||{};window.LaravelDataTables["dataTableBuilder"]=$("#dataTableBuilder").DataTable({"serverSide":true,"processing":true,"ajax":{"url":"","type":"GET","data":function(data) {
            for (var i = 0, len = data.columns.length; i < len; i++) { 
                if (!data.columns[i].search.value) delete data.columns[i].search;
                if (data.columns[i].searchable === true) delete data.columns[i].searchable;
                if (data.columns[i].orderable === true) delete data.columns[i].orderable;
                if (data.columns[i].data === data.columns[i].name) delete data.columns[i].name;
            }
            delete data.search.regex;}},"columns":[{"name":"id","data":"id","title":"Id","orderable":true,"searchable":true},{"name":"username","data":"username","title":"Username","orderable":true,"searchable":true},{"name":"name","data":"name","title":"Name","orderable":true,"searchable":true},{"name":"email","data":"email","title":"Email","orderable":true,"searchable":true},{"name":"telephonenumber","data":"telephonenumber","title":"Telephonenumber","orderable":true,"searchable":true},{"name":"displayname","data":"displayname","title":"Displayname","orderable":true,"searchable":true},{"name":"created_at","data":"created_at","title":"Created At","orderable":true,"searchable":true},{"name":"updated_at","data":"updated_at","title":"Updated At","orderable":true,"searchable":true},{"defaultContent":"","data":"action","name":"action","title":"Action","render":null,"orderable":false,"searchable":false,"width":"80px"}],"dom":"Bfrtip","order":[[0,"desc"]],"buttons":["create","export","print","reset","reload"]});})(window,jQuery);
</script>

</head>
<body>



    <div id="app">
        

        

            <table  class="table" id="dataTableBuilder"><thead><tr><th >Id</th><th >Username</th><th >Name</th><th >Email</th><th >Telephonenumber</th><th >Displayname</th><th >Created At</th><th >Updated At</th><th  width="80px">Action</th></tr></thead></table>
    </div>

    <!-- Scripts -->
    <script src="https://dash.app/js/app.js"></script>
</body>
</html>
```

**Comment by jamesriady** [comment]
what template do you use? is that adminlte?
if you are using adminlte

change:
`@push('scripts')
    <link rel="stylesheet" href="https://cdn.datatables.net/buttons/1.0.3/css/buttons.dataTables.min.css">
    <script src="https://cdn.datatables.net/buttons/1.0.3/js/dataTables.buttons.min.js"></script>
    <script src="/vendor/datatables/buttons.server-side.js"></script>
    {!! $dataTable->scripts() !!}
@endpush`

into
`@section('scripts')`
`@parent`
`<link href="//netdna.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">`
`<link rel="stylesheet" href="//cdn.datatables.net/1.10.7/css/jquery.dataTables.min.css">`
`<script src="//cdn.datatables.net/1.10.7/js/jquery.dataTables.min.js"></script>`
`{!! $dataTable->scripts() !!}`
`@endsection`

**Comment by swerwajod** [comment]
@jameriady It's not adminlte just default Laravel. However, I tried the content you suggested with no change.  The Push is successfully placing the content in the <Head> area. 
@yajra I also changed the cdn info with that build from the link you provided. Again still a header row only.

**Comment by swerwajod** [comment]
Ok, I cloned your dt54 demo and checked that I had all the scripts and such to see what was missing and to get the scripts in the right order.  So the first part of the Issue was js order. I used the link above to get the datatable js links. and the @stack('scripts') needed to be just before my app.js. 

That aside I now have what looks to be a properly formatted data table header with icons instead of just column names. So it appears that the js is working.

![image](https://user-images.githubusercontent.com/32367723/31022488-f1312e38-a506-11e7-91a6-35d52f7baf8b.png)

However, you can see it saw processing continually with no data.  An error comes up and states.

![image](https://user-images.githubusercontent.com/32367723/31022612-4c1f555e-a507-11e7-8f11-f695b4d8cbb4.png)

In the console, it appears that there is no JSON from datatables. Still researching.

**Comment by jamesriady** [comment]
You said that you are using datatables service as implementation right..
Let's try this, change in your userDatatable..
`public function query(User $model)
    {
        $query = User::query();
        return $this->applyScopes($query);
    }`

**Comment by swerwajod** [comment]
Sorry I found a typo in that method that I made with all the changing around.  So now in the console, I do see the correct response with the original method posted.  I also used the method you posted above and I get a correct response as well. (in the console).  
The data table still empty so I think it must be something I'm still not including from the .js or in the wrong order.

**Comment by swerwajod** [comment]
OK, I got it working. When looking through the dt54 demo I noticed an additional settings.js file in the resources/assets/js/ folder.  I included it in the app.js file like in the demo. 
After I running the npm run dev to remix the published css and js, it worked.
It may be obvious to others what I missed but perhaps adding a page to you documentation that explains the required parts of jquery, the page script needed to make it execute and the order in which it needs to be loaded on the page may be very helpful to many. Or just an explanation here would be helpful. Thanks for this package.
